## **Import Library**

In [47]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

## **Gathering Dataset**

In [48]:
inshot = pd.read_csv("https://raw.githubusercontent.com/yovelakalista23/BPML/refs/heads/main/Proyek-Analisis-Sentimen/ulasan_inshot.csv")
inshot.head()

,reviewId,userName,userImage,content,score,thumbsUpCount,reviewCreatedVersion,at,replyContent,repliedAt,appVersion,text_clean,text_caseFoldingText,text_slangwords,text_tokenizingText,text_stopword,text_akhir,polarity_score,polarity
0,3bf58ad8-c392-47aa-8096-8c929a2371f6,Pengguna Google,https://play-lh.googleusercontent.com/EGemoI2N...,"Lumayan membantu para kreator pemula, yg pengh...",5,3,2.123.1487,2025-05-12 04:38:04,Unknown,Unknown,2.123.1487,Lumayan membantu para kreator pemula yg pengha...,lumayan membantu para kreator pemula yg pengha...,lumayan membantu para kreator pemula yang peng...,"['lumayan', 'membantu', 'para', 'kreator', 'pe...","['lumayan', 'membantu', 'kreator', 'pemula', '...",lumayan membantu kreator pemula penghasilannya...,16,positive
1,20e4333b-7877-4135-b01b-ea0346d8ee94,Pengguna Google,https://play-lh.googleusercontent.com/EGemoI2N...,aplikasi edit ini bagus untuk saya yg pemula d...,5,123,2.115.1483,2025-04-18 21:57:20,Unknown,Unknown,2.115.1483,aplikasi edit ini bagus untuk saya yg pemula d...,aplikasi edit ini bagus untuk saya yg pemula d...,aplikasi edit ini bagus untuk saya yang pemula...,"['aplikasi', 'edit', 'ini', 'bagus', 'untuk', ...","['aplikasi', 'edit', 'bagus', 'pemula', 'editi...",aplikasi edit bagus pemula editing sederhana a...,4,positive
2,38c1810c-e192-446f-a7c4-80c02de7b905,Pengguna Google,https://play-lh.googleusercontent.com/EGemoI2N...,Aplikasinya bagus banget.. fitur gratisnya pun...,5,138,2.115.1483,2025-04-25 11:55:26,Unknown,Unknown,2.115.1483,Aplikasinya bagus banget fitur gratisnya pun j...,aplikasinya bagus banget fitur gratisnya pun j...,aplikasinya bagus banget fitur gratisnya pun j...,"['aplikasinya', 'bagus', 'banget', 'fitur', 'g...","['aplikasinya', 'bagus', 'banget', 'fitur', 'g...",aplikasinya bagus banget fitur gratisnya sesua...,20,positive
3,e1067cc6-dd79-4709-a621-c6d0c94c3adf,Pengguna Google,https://play-lh.googleusercontent.com/EGemoI2N...,"kasih bintang 4 dulu karena mudah editnya, dan...",4,0,2.123.1487,2025-05-09 17:55:44,Unknown,Unknown,2.123.1487,kasih bintang dulu karena mudah editnya dan g...,kasih bintang dulu karena mudah editnya dan g...,kasih bintang dulu karena mudah editnya dan gr...,"['kasih', 'bintang', 'dulu', 'karena', 'mudah'...","['kasih', 'bintang', 'mudah', 'editnya', 'grat...",kasih bintang mudah editnya gratisan edit vide...,11,positive
4,8391b32f-490e-4df9-a7fb-c318813155b5,Pengguna Google,https://play-lh.googleusercontent.com/EGemoI2N...,aplikasi ini dari dulu selalu konsisten member...,5,68,2.115.1483,2025-04-22 18:53:21,Unknown,Unknown,2.115.1483,aplikasi ini dari dulu selalu konsisten member...,aplikasi ini dari dulu selalu konsisten member...,aplikasi ini dari dulu selalu konsisten member...,"['aplikasi', 'ini', 'dari', 'dulu', 'selalu', ...","['aplikasi', 'konsisten', 'pelayanan', 'fitur'...",aplikasi konsisten pelayanan fitur fitur berba...,0,neutral


## **Assessing Dataset**

In [49]:
inshot.shape

(45000, 19)

In [50]:
inshot.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45000 entries, 0 to 44999
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   reviewId              45000 non-null  object
 1   userName              45000 non-null  object
 2   userImage             45000 non-null  object
 3   content               45000 non-null  object
 4   score                 45000 non-null  int64 
 5   thumbsUpCount         45000 non-null  int64 
 6   reviewCreatedVersion  45000 non-null  object
 7   at                    45000 non-null  object
 8   replyContent          45000 non-null  object
 9   repliedAt             45000 non-null  object
 10  appVersion            45000 non-null  object
 11  text_clean            44914 non-null  object
 12  text_caseFoldingText  44914 non-null  object
 13  text_slangwords       44914 non-null  object
 14  text_tokenizingText   45000 non-null  object
 15  text_stopword         45000 non-null

In [51]:
inshot.isnull().sum()

,0
reviewId,0
userName,0
userImage,0
content,0
score,0
thumbsUpCount,0
reviewCreatedVersion,0
at,0
replyContent,0
repliedAt,0


In [52]:
print("Number of Duplicates: ", inshot.duplicated().sum())

Number of Duplicates:  0


In [53]:
inshot.describe()

,score,thumbsUpCount,polarity_score
count,45000.000000,45000.000000,45000.000000
mean,4.276711,6.761311,2.752422
std,1.279462,73.472786,6.374992
min,1.000000,0.000000,-68.000000
25%,4.000000,0.000000,-1.000000
50%,5.000000,0.000000,3.000000
75%,5.000000,0.000000,6.000000
max,5.000000,3452.000000,49.000000


## **CLeaning Dataset**

In [54]:
inshot = inshot.dropna()

In [55]:
inshot.info()

<class 'pandas.core.frame.DataFrame'>
Index: 44904 entries, 0 to 44999
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   reviewId              44904 non-null  object
 1   userName              44904 non-null  object
 2   userImage             44904 non-null  object
 3   content               44904 non-null  object
 4   score                 44904 non-null  int64 
 5   thumbsUpCount         44904 non-null  int64 
 6   reviewCreatedVersion  44904 non-null  object
 7   at                    44904 non-null  object
 8   replyContent          44904 non-null  object
 9   repliedAt             44904 non-null  object
 10  appVersion            44904 non-null  object
 11  text_clean            44904 non-null  object
 12  text_caseFoldingText  44904 non-null  object
 13  text_slangwords       44904 non-null  object
 14  text_tokenizingText   44904 non-null  object
 15  text_stopword         44904 non-null  obj

## **Splitting**

In [56]:
# Memisahkan kolom 'statement' menjadi fitur (X) dan kolom 'status' menjadi target (y)
X = inshot['text_akhir']
y = inshot['polarity']

### **Ekstraksi Fitur TF-IDF**

In [58]:
# Inisialisasi TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer(
    max_features=1000,      # Hanya mengambil 1000 fitur terpenting
    min_df=5,               # Hanya muncul minimal di 5 dokumen
    max_df=0.85,            # Tidak lebih dari 85% dokumen
    ngram_range=(1,2)       # Menggunakan unigram dan bigram
)

# Transformasi menjadi matriks TF-IDF
tfidf_matrix = tfidf_vectorizer.fit_transform(X)

tfidf_matrix

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 316623 stored elements and shape (44904, 1000)>

### **Ekstraksi fitur BoW**

In [59]:
# Inisialisasi CountVectorizer untuk Bag of Words
bow_vectorizer = CountVectorizer(
    max_features=1000,      # Hanya 1000 fitur terpenting
    min_df=5,               # Hanya muncul minimal di 5 dokumen
    max_df=0.85,            # Tidak lebih dari 85% dokumen
    ngram_range=(1,2)       # Menggunakan unigram dan bigram
)

# Transformasi menjadi matriks BoW
bow_matrix = bow_vectorizer.fit_transform(X)

# Konversi hasil menjadi DataFrame
X_bow = pd.DataFrame(bow_matrix.toarray(), columns=bow_vectorizer.get_feature_names_out())

# Menampilkan hasil BoW
X_bow.head()

,ad,admin,aj,ajah,ajh,ak,akun,alhamdulillah,allah,ama,...,wkwk,wow,wr,wr wb,wrwb,yahh,yh,yng,youtube,youtuber
0,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [60]:
# 80/20 TF-IDF
X_train1, X_test1, y_train1, y_test1 = train_test_split(tfidf_matrix, y, test_size=0.2, random_state=42)

# 70/30 TF-IDF
X_train2, X_test2, y_train2, y_test2 = train_test_split(tfidf_matrix, y, test_size=0.3, random_state=42)

# 80/20 BoW
X_train3, X_test3, y_train3, y_test3 = train_test_split(X_bow, y, test_size=0.2, random_state=42)

## **Modeling**

In [67]:
# Model SVM (Support Vector Machine)
svm_model = SVC(kernel='linear', random_state=42)
# Training dan Evaluasi Model 1: SVM dengan TF-IDF (80/20)
svm_model.fit(X_train1, y_train1)

y_train_pred1 = svm_model.predict(X_train1)
y_test_pred1 = svm_model.predict(X_test1)

hasil1 = {
    'Model': 'SVM TF-IDF(80/20)',
    'Train Accuracy': accuracy_score(y_train1, y_train_pred1),
    'Test Accuracy': accuracy_score(y_test1, y_test_pred1)
}

In [68]:
# Model Random Forest (RF)
rf_modeltfidf = RandomForestClassifier(n_estimators=100, random_state=42)
# Training dan Evaluasi Model 2: RF dengan TF-IDF (70/30)
rf_modeltfidf.fit(X_train2, y_train2)

y_train_pred2 = rf_modeltfidf.predict(X_train2)
y_test_pred2 = rf_modeltfidf.predict(X_test2)

hasil2 = {
    'Model': 'Random Forest TF-IDF(70/30)',
    'Train Accuracy': accuracy_score(y_train2, y_train_pred2),
    'Test Accuracy': accuracy_score(y_test2, y_test_pred2)
}

In [69]:
# Model Random Forest (RF)
rf_modelbow = RandomForestClassifier(n_estimators=100, random_state=42)
# Training dan Evaluasi Model 3: RF dengan BoW (80/20)
rf_modelbow.fit(X_train3, y_train3)

y_train_pred3 = rf_modelbow.predict(X_train3)
y_test_pred3 = rf_modelbow.predict(X_test3)

hasil3 = {
    'Model': 'Random Forest BoW(80/20)',
    'Train Accuracy': accuracy_score(y_train3, y_train_pred3),
    'Test Accuracy': accuracy_score(y_test3, y_test_pred3)
}

In [70]:
# Menampilkan hasil dalam bentuk dataframe
results = [hasil1, hasil2, hasil3]
accuracy_df = pd.DataFrame(results)
print(accuracy_df)

                         Model  Train Accuracy  Test Accuracy
0            SVM TF-IDF(80/20)        0.929460       0.916602
1  Random Forest TF-IDF(70/30)        0.992492       0.864163
2     Random Forest BoW(80/20)        0.992512       0.865160


Menggunakan 3 skema pelatihan dan 2 kombinasi, yaitu

- SVM (Support Vector Machine) menggunakan TF-IDF mendapatkan hasil train akurasi 92.9% dan test akurasinya 91.6%
- Random Forest menggunakan TF-IDF mendapatkan hasil train akurasi 99.2% dan test akurasi 86.4%
- Random Forest menggunakan BoW (Bag of Words) mendapatkan hasil train akurasi 99.2% dan test akurasi 86.5%

Dari ketiga skema tersebut, SVM mendapatkan hasil akurasi yang sangat baik untuk train dan test akurasinya, tetapi Random Forest TD-IDF dan BoW juga mendapatkan hasil akurasi yang cukup baik.

## **Testing**

In [71]:
# Kalimat untuk testing
test_samples = [
    "Produk ini sangat buruk, saya kecewa!",
    "Barang sesuai dengan deskripsi, kualitas bagus.",
    "Pengiriman lambat, tapi produk oke."
]

# Transform teks menjadi TF-IDF
test_samples_tfidf = tfidf_vectorizer.transform(test_samples)

# Melakukan inference menggunakan model SVM
svm_predictions = svm_model.predict(test_samples_tfidf)

# Melakukan inference menggunakan model Random Forest (TF-IDF 70/30)
rf_tfidf_predictions = rf_modeltfidf.predict(test_samples_tfidf)

# Melakukan inference menggunakan model Random Forest (TF-IDF 80/20)
rf_bow_predictions = rf_modelbow.predict(test_samples_tfidf)

# Membuat dataframe hasil inference
inference_df = pd.DataFrame({
    'Text': test_samples,
    'SVM Prediction': svm_predictions,
    'Random Forest TF-IDF': rf_tfidf_predictions,
    'Random Forest BoW': rf_bow_predictions
})

# Menampilkan hasil inference
print(inference_df)

                                              Text SVM Prediction  \
0            Produk ini sangat buruk, saya kecewa!       negative   
1  Barang sesuai dengan deskripsi, kualitas bagus.       positive   
2              Pengiriman lambat, tapi produk oke.       positive   

  Random Forest TF-IDF Random Forest BoW  
0             negative          negative  
1             positive          positive  
2             positive          positive  


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


Setelah melakukan testing/inference dengan text sample, hasil dari 3 skema tersebut sesuai dan sama.